contributors/design-proposals/scheduling/scheduling-framework.md

```diff
@@ -1,436 +1,618 @@
-These plugins are called serially.
+This is an informational extension point. Plugins can will be called with a list
+of nodes that passed the filtering phase. A plugin may use this data to update
+internal state or to generate logs/metrics.
+
+**Note:** Plugins wishing to perform "pre-scoring" work should use the
+post-filter extension point.
 
 ### Scoring
 
-These plugins are similar to priority function in scheduler v1. They are
-utilized to rank nodes that have passed the filtering stage. Similar to Filter
-plugins, these are called per node serially in the same order registered, but
-scheduler may run them for multiple nodes in parallel.  
-Each one of these functions return a score for the given node. The score is
-multiplied by the weight of the function and aggregated with the result of other
-scoring functions to yield a total score for the node.  
-These functions can never block scheduling. In case of an error they should
-return zero for the Node being ranked.
+These plugins are used to rank nodes that have passed the filtering phase. The
+scheduler will call each scoring plugin for each node. There will be a well
+defined range of integers representing the minimum and maximum scores. After the
+[normalize scoring](#normalize-scoring) phase, the scheduler will combine node
+scores from all plugins according to the configured plugin weights.
 
-### Post-scoring/pre-reservation
+If a scoring plugin returns an error, the scheduler will treat it as a zero
+score.
 
-After all scoring plugins are invoked and the score of nodes are determined, the
-framework picks the best node with the highest score and then it calls
-post-scoring plugins. The Pod and the chosen Node are passed to these plugins.
-These plugins have one more chance to check any conditions about the assignment
-of the Pod to this Node and reject the node if needed.
+### Normalize scoring
 
-![image](images/scheduling-framework-threads.png)
+These plugins are used to modify scores before the scheduler computes a final
+ranking of Nodes. A plugin that registers for this extension point will be
+called with the [scoring](#scoring) results from the same plugin. This is called
+once per plugin per scheduling cycle.
+
+For example, suppose a plugin `BlinkingLightScorer` ranks Nodes based on how
+many blinking lights they have.
+
+```go
+func ScoreNode(_ *v1.Pod, n *v1.Node) (int, error) {
+   return getBlinkingLightCount(n)
+}
+```
+
+However, the maximum count of blinking lights may be small compared to
+`NodeScoreMax`. To fix this, `BlinkingLightScorer` should also register for this
+extension point.
+
+```go
+func NormalizeScores(scores map[string]int) {
+   highest := 0
+   for _, score := range scores {
+      highest = max(highest, score)
+   }
+   for node, score := range scores {
+      scores[node] = scores*NodeScoreMax/highest
+   }
+}
+```
+
+If any post-scoring plugin returns an error, the scheduling cycle is aborted.
```

Comment:
Should we standardize this extension point's name? We used `Normalize scoring` in title, but `post-scoring` in the content.

Reply:
That was my mistake. Good catch.